standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java

```diff
@@ -1104,89 +1104,101 @@ public boolean checkFailedCompactions(CompactionInfo ci) throws MetaException {
+            pStmt.setLong(1, ci.id);
+            LOG.debug("Going to execute update <" + s + ">");
+            int updCnt = pStmt.executeUpdate();
+          } else {
+            if (ci.id > 0) {
+              //the record with valid CQ_ID has disappeared - this is a sign of something wrong
+              throw new IllegalStateException("No record with CQ_ID=" + ci.id + " found in COMPACTION_QUEUE");
+            }
+          }
+
+          if (ci.id == 0) {
+            //The failure occurred before we even made an entry in COMPACTION_QUEUE
+            //generate ID so that we can make an entry in COMPLETED_COMPACTIONS
+            ci.id = generateCompactionQueueId(dbConn);
+            //mostly this indicates that the Initiator is paying attention to some table even though
+            //compactions are not happening.
+            ci.state = DID_NOT_INITIATE;
+            //this is not strictly accurate, but 'type' cannot be null.
+            if (ci.type == null) {
+              ci.type = CompactionType.MINOR;
+            }
+            ci.start = getDbTime(dbConn);
+            LOG.debug("The failure occurred before we even made an entry in COMPACTION_QUEUE. Generated ID so that we "
+                    + "can make an entry in COMPLETED_COMPACTIONS. New Id: " + ci.id);
+          } else {
+            ci.state = FAILED_STATE;
+          }
+
+          if(errorMessage != null) {
+            ci.errorMessage = errorMessage;
+          }
+
+          close(rs, pStmt, null);
         }
-        close(rs, stmt, null);
-        closeStmt(pStmt);
 
         pStmt = dbConn.prepareStatement("INSERT INTO \"COMPLETED_COMPACTIONS\" "
             + "(\"CC_ID\", \"CC_DATABASE\", \"CC_TABLE\", \"CC_PARTITION\", \"CC_STATE\", \"CC_TYPE\", "
             + "\"CC_TBLPROPERTIES\", \"CC_WORKER_ID\", \"CC_START\", \"CC_END\", \"CC_RUN_AS\", "
             + "\"CC_HIGHEST_WRITE_ID\", \"CC_META_INFO\", \"CC_HADOOP_JOB_ID\", \"CC_ERROR_MESSAGE\", "
             + "\"CC_ENQUEUE_TIME\", \"CC_WORKER_VERSION\", \"CC_INITIATOR_ID\", \"CC_INITIATOR_VERSION\") "
             + "VALUES(?,?,?,?,?, ?,?,?,?,?, ?,?,?,?,?,?,?,?,?)");
-        if (errorMessage != null) {
-          ci.errorMessage = errorMessage;
-        }
+
         CompactionInfo.insertIntoCompletedCompactions(pStmt, ci, getDbTime(dbConn));
         int updCount = pStmt.executeUpdate();
         LOG.debug("Inserted " + updCount + " entries into COMPLETED_COMPACTIONS");
         LOG.debug("Going to commit");
         closeStmt(pStmt);
         dbConn.commit();
       } catch (SQLException e) {
-        LOG.warn("markFailed(" + ci.id + "):" + e.getMessage());
+        LOG.warn("markFailed(" + ((ci != null) ? ci.id : "missing id") + "):" + e.getMessage());
         LOG.debug("Going to rollback");
         rollbackDBConn(dbConn);
         checkRetryable(dbConn, e, "markFailed(" + ci + ")");
         LOG.error("markFailed(" + ci + ") failed: " + e.getMessage(), e);
```

Comment:
These will call ci#toString, right?

Reply:
Not really.
If the `ci == null` then it just adds nullm otherwise the `toString()` getting triggered.
```java
    CompactionInfo ci = null;
    System.out.println("(" + ci + ")");

    ci = new CompactionInfo("x", "y", "z", CompactionType.MINOR);
    System.out.println("(" + ci + ")");
```
```
(null)
(id:0,dbname:x,...)
```

Reply:
Got it, thanks!